improver/calibration/reliability_calibration.py

```diff
@@ -0,0 +1,414 @@
+        try:
+            n_forecast_periods, = forecasts.coord('forecast_period').shape
+        except CoordinateNotFoundError:
+            n_forecast_periods = 0
+        if n_cycle_hours != 1 or n_forecast_periods != 1:
+            msg = ('Forecasts have been provided from differing cycle hours '
+                   'or forecast periods, or without these coordinates. These '
+                   'coordinates should be present and consistent between '
+                   'forecasts. Number of cycle hours found: {}, number of '
+                   'forecast periods found: {}.')
+            raise ValueError(msg.format(n_cycle_hours, n_forecast_periods))
+
+    def _create_cycle_hour_coord(self, forecast_reference_time):
+        """
+        Constructs a coordinate that contains the cycle hour for which the
+        reliability table is valid.
+
+        Args:
+            forecast_reference_time (iris.coord.DimCoord):
+                The forecast_reference_time coordinate to be uses in the
+                coordinate creation.
+        Returns:
+            iris.coord.DimCoord:
+                A dimension coordinate containing an integer unitless
+                representation of the cycle hour.
+        """
+        cycle_hour, = self._get_cycle_hours(forecast_reference_time)
+        cycle_coord = iris.coords.DimCoord(cycle_hour, long_name='cycle_hour',
+                                           units=1)
+        return cycle_coord
+
+    @staticmethod
+    def _define_metadata(forecast_slice, diagnostic):
+        """
+        Define metadata that is specifically required for reliability table
+        cubes, whilst ensuring any mandatory attributes are also populated.
+
+        Args:
+            forecast_slice (iris.cube.Cube):
+                The source cube from which to get pre-existing metadata of use.
+            diagnostic (str):
+                The name of the diagnostic within the cube.
+        Returns:
+            dict:
+                A dictionary of attributes that are appropriate for the
+                reliability table cube.
+        """
+        attributes = generate_mandatory_attributes([forecast_slice])
+        attributes["title"] = "Reliability calibration data table"
+        attributes["diagnostic_standard_name"] = diagnostic
+        return attributes
+
+    def _create_reliability_table_cube(self, forecast, threshold_coord):
+        """
+        Construct a reliability table cube and populate it with the provided
+        data.
+
+        Args:
+            forecast (iris.cube.Cube):
+                A cube slice across the spatial dimensions of the forecast
+                data. This slice provides the time and threshold values that
+                relate to the reliability_table_data.
+            threshold_coord (iris.coords.DimCoord):
+                The threshold coordinate.
+        Returns:
+            iris.cube.Cube:
+                A reliability table cube.
```

Comment:
Somewhere I think we should add some extended documentation showing an example of a reliability table. I don't know if it needs to go here or in the process docstring. Maybe the cube example is best in process and here it should just say what the coordinates on the reliability table cube should be and what order they will be in.

Reply:
I've slightly expanded the metadata doc-string. I've added an example reliability table cube to the process doc-string as an extended-documentation add on and described its structure a little.